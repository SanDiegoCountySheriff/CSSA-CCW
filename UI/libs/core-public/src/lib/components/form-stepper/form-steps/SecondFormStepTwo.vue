<template>
  <div>
    <DocumentInfoSection />
    <v-form
      ref="form"
      v-model="state.valid"
    >
      <v-subheader class="sub-header font-weight-bold">
        {{ $t('File Upload') }}
      </v-subheader>
      <v-row>
        <v-col
          cols="12"
          lg="6"
        >
          <v-file-input
            ref="driver-license"
            show-size
            small-chips
            accept="image/png, image/jpeg"
            :label="$t('Driver License')"
            @change="handleFileInput($event, 'DriverLicense')"
          />
        </v-col>
        <v-col
          cols="12"
          lg="6"
        >
          <v-file-input
            show-size
            small-chips
            accept=".pdf, .doc, .docx"
            :label="$t('Proof of Residence 1')"
            @change="handleFileInput($event, 'ProofResidency')"
          />
        </v-col>
        <v-col
          cols="12"
          lg="6"
        >
          <v-file-input
            show-size
            small-chips
            accept=".pdf, .doc, .docx"
            :label="$t('Proof of Residence 2')"
            @change="handleFileInput($event, 'ProofResidency2')"
          />
        </v-col>
      </v-row>
      <v-divider />
      <v-subheader class="sub-header font-weight-bold">
        {{ $t(' Military') }}
      </v-subheader>
      <v-row>
        <v-col
          cols="12"
          lg="6"
        >
          <v-file-input
            show-size
            small-chips
            accept=".pdf, .doc, .docx"
            :label="$t('Military Document')"
            @change="handleFileInput($event, 'MilitaryDoc')"
          />
        </v-col>
      </v-row>
      <v-divider />
      <v-subheader class="sub-header font-weight-bold">
        {{ $t(' Not born in US') }}
      </v-subheader>
      <v-row>
        <v-col
          cols="12"
          lg="6"
        >
          <v-file-input
            show-size
            small-chips
            accept=".pdf, .doc, .dox"
            :label="$t('Citizenship Documents')"
            @change="handleFileInput($event, 'Citizenship')"
          />
        </v-col>
      </v-row>

      <v-divider />
      <v-subheader class="sub-header font-weight-bold">
        {{ $t(' Supporting Documents') }}
      </v-subheader>
      <v-row>
        <v-col
          cols="12"
          lg="6"
        >
          <v-file-input
            show-size
            small-chips
            multiple
            :label="$t('Supporting Documents')"
            @change="handleFileInput($event, 'Supporting')"
          />
        </v-col>
      </v-row>
      <v-divider />

      <v-subheader class="sub-header font-weight-bold">
        {{ $t(' Legal name change') }}
      </v-subheader>
      <v-row>
        <v-col
          cols="12"
          lg="6"
        >
          <v-file-input
            show-size
            small-chips
            accept=".pdf, .doc, .dox"
            :label="$t('Name change documents')"
            @change="handleFileInput($event, 'NameChange')"
          />
        </v-col>
      </v-row>
      <v-divider />
      <v-subheader class="sub-header font-weight-bold">
        {{ $t(' License Type Documents') }}
      </v-subheader>
      <v-row>
        <v-col
          cols="12"
          lg="6"
        >
          <v-file-input
            show-size
            small-chips
            accept=".pdf, .doc, .dox"
            :label="$t('Judicial documents')"
            @change="handleFileInput($event, 'Judicial')"
          />
        </v-col>
        <v-col
          cols="12"
          lg="6"
        >
          <v-file-input
            show-size
            small-chips
            accept=".pdf, .doc, .dox"
            :label="$t('Reserve documents')"
            @change="handleFileInput($event, 'Reserve')"
          />
        </v-col>
      </v-row>
      <v-divider />
    </v-form>
    <v-progress-circular
      v-if="!state.uploadSuccessful"
      color="primary"
      indeterminate
    />
    <FormButtonContainer
      v-else
      :valid="state.valid"
      @submit="handleSubmit"
      @back="handlePreviousSection"
      @cancel="router.push('/')"
    />
    <v-snackbar
      v-model="state.snackbar"
      :timeout="2000"
      bottom
      color="error"
      outlined
    >
      {{ $t('Section update unsuccessful please try again.') }}
    </v-snackbar>
  </div>
</template>

<script setup lang="ts">
import DocumentInfoSection from '@shared-ui/components/info-sections/DocumentInfoSection.vue';
import Endpoints from '@shared-ui/api/endpoints';
import FormButtonContainer from '@core-public/components/containers/FormButtonContainer.vue';
import { UploadedDocType } from '@shared-utils/types/defaultTypes';
import axios from 'axios';
import { reactive } from 'vue';
import { useCompleteApplicationStore } from '@core-public/stores/completeApplication';
import { useMutation } from '@tanstack/vue-query';
import { useRouter } from 'vue-router/composables';

const applicationStore = useCompleteApplicationStore();
const completeApplication = applicationStore.completeApplication.application;
const router = useRouter();

interface ISecondFormStepTwoProps {
  handleNextSection: CallableFunction;
  handlePreviousSection: CallableFunction;
}

const props = defineProps<ISecondFormStepTwoProps>();

const state = reactive({
  driver: {} as File,
  files: [] as Array<{ form; target }>,
  valid: false,
  uploadSuccessful: true,
  snackbar: false,
});

const fileMutation = useMutation({
  mutationFn: handleFileUpload,
  onSuccess: () => {
    state.uploadSuccessful = true;
    completeApplication.currentStep = 9;
    props.handleNextSection();
  },
  onError: () => {
    state.snackbar = true;
  },
});

function handleFileInput(event: File, target: string) {
  const form = new FormData();

  form.append('fileToPersist', event);

  const fileObject = {
    form,
    target,
  };

  state.files.push(fileObject);
  window.console.log(state.files);
}

async function handleFileUpload() {
  state.files.forEach(file => {
    const newFileName = `${applicationStore.completeApplication.application.orderId}_${completeApplication.personalInfo.lastName}_${completeApplication.personalInfo.firstName}_${file.target}`;

    axios
      .post(
        `${Endpoints.POST_DOCUMENT_IMAGE_ENDPOINT}?saveAsFileName=${newFileName}`,
        file.form
      )
      .catch(e => {
        window.console.warn(e);
        Promise.reject();
      });

    const uploadDoc: UploadedDocType = {
      DocumentType: file.target,
      name: `${newFileName}`,
      uploadedBy: completeApplication.userEmail,
      uploadedDateTimeUtc: new Date(Date.now()).toISOString(),
    };

    completeApplication.uploadedDocuments.push(uploadDoc);
  });
}

function handleSubmit() {
  state.uploadSuccessful = false;
  fileMutation.mutate();
}
</script>

<style lang="scss" scoped>
.text-line {
  display: flex;
  flex-direction: row;
}
</style>
