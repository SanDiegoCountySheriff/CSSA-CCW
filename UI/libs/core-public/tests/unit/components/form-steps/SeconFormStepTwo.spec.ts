import SecondFormStepTwo from '@core-public/../../../../../shared/ui/src/lib/components/form-stepper/form-steps/SecondFormStepTwo.vue';
import Vuetify from 'vuetify';
import { createLocalVue, mount } from '@vue/test-utils';
import { createTestingPinia } from '@pinia/testing';

const pinia = createTestingPinia();
const localVue = createLocalVue();
const tMock = {
  $t: text => text,
};

describe('SecondFormStepTwo', () => {
  let vuetify;
  let wrapper;

  beforeEach(() => {
    vuetify = new Vuetify();
    wrapper = mount(SecondFormStepTwo, {
      localVue,
      vuetify,
      pinia,
      mocks: tMock,
      propsData: {
        handleNextSection: () => null,
      },
    });
  });
  it('Should match the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });
});
