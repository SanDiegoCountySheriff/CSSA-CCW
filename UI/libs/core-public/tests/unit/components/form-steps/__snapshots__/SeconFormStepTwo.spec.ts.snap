// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SecondFormStepTwo Should match the snapshot 1`] = `
<div>
  <div class="text-left v-card v-sheet theme--light">
    <div class="v-card__title"> Documentation-Header</div>
    <p class="ml-5">
      Documentation-Subheader
    </p>
    <div class="v-card__text">
      <div class="text-line">
        <h3 class="text-label">
          EVERYONE
        </h3>
        <p class="text-info">
          Documentation-Everyone
        </p>
      </div>
      <div class="text-line">
        <h3 class="text-label">
          MILITARY
        </h3>
        <p class="text-info">
          Documentation-Military
        </p>
      </div>
      <div class="text-line">
        <h3 class="text-label">
          NOT-BORN-IN-US
        </h3>
        <p class="text-info">
          Documentation-NotBornInUs
        </p>
      </div>
      <div class="mb-3">
        <ul>
          <li>Documentation-bullet1</li>
          <li>Documentation-bullet2</li>
          <li>Documentation-bullet3</li>
          <li>Documentation-bullet4</li>
        </ul>
      </div>
      <div class="text-line">
        <h3 class="text-label">
          SUPPORTING
        </h3>
        <p class="text-info">
          Documentation-Supporting
        </p>
      </div>
      <div class="text-line">
        <h3 class="text-label">
          LEGAL-NAME-CHANGE
        </h3>
        <p class="text-info">
          Documentation-LegalNameChange
        </p>
      </div>
      <div class="text-line">
        <h3 class="text-label">
          JUDICIAL
        </h3>
        <p class="text-info">
          Documentation-Judicial
        </p>
      </div>
      <div class="text-line">
        <h3 class="text-label">
          RESERVE
        </h3>
        <p class="text-info">
          Documentation-Reserve
        </p>
      </div>
    </div>
  </div>
  <form novalidate="novalidate" class="v-form">
    <div class="v-subheader sub-header font-weight-bold theme--light">
      File Upload
    </div>
    <div class="row">
      <div class="col-sm-1 col-lg-4 col-4">
        <div class="v-input theme--light v-text-field v-file-input">
          <div class="v-input__prepend-outer">
            <div class="v-input__icon v-input__icon--prepend"><button type="button" aria-label="prepend icon" class="v-icon notranslate v-icon--link mdi mdi-paperclip theme--light"></button></div>
          </div>
          <div class="v-input__control">
            <div class="v-input__slot">
              <div class="v-text-field__slot"><label for="input-6" class="v-label theme--light" style="left: 0px; position: absolute;">Driver License</label>
                <div class="v-file-input__text v-file-input__text--chips"></div><input accept="image/png, image/jpeg" id="input-6" type="file">
              </div>
              <div class="v-input__append-inner">
                <div></div>
              </div>
            </div>
            <div class="v-text-field__details">
              <div class="v-messages theme--light">
                <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div class="col-sm-1 col-lg-4 col-4">
        <div class="v-input theme--light v-text-field v-file-input">
          <div class="v-input__prepend-outer">
            <div class="v-input__icon v-input__icon--prepend"><button type="button" aria-label="prepend icon" class="v-icon notranslate v-icon--link mdi mdi-paperclip theme--light"></button></div>
          </div>
          <div class="v-input__control">
            <div class="v-input__slot">
              <div class="v-text-field__slot"><label for="input-10" class="v-label theme--light" style="left: 0px; position: absolute;">Proof of Residence 1</label>
                <div class="v-file-input__text v-file-input__text--chips"></div><input accept=".pdf, .doc, .docx" id="input-10" type="file">
              </div>
              <div class="v-input__append-inner">
                <div></div>
              </div>
            </div>
            <div class="v-text-field__details">
              <div class="v-messages theme--light">
                <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div class="col-sm-1 col-lg-4 col-4">
        <div class="v-input theme--light v-text-field v-file-input">
          <div class="v-input__prepend-outer">
            <div class="v-input__icon v-input__icon--prepend"><button type="button" aria-label="prepend icon" class="v-icon notranslate v-icon--link mdi mdi-paperclip theme--light"></button></div>
          </div>
          <div class="v-input__control">
            <div class="v-input__slot">
              <div class="v-text-field__slot"><label for="input-14" class="v-label theme--light" style="left: 0px; position: absolute;">Proof of Residence 2</label>
                <div class="v-file-input__text v-file-input__text--chips"></div><input accept=".pdf, .doc, .docx" id="input-14" type="file">
              </div>
              <div class="v-input__append-inner">
                <div></div>
              </div>
            </div>
            <div class="v-text-field__details">
              <div class="v-messages theme--light">
                <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <hr role="separator" aria-orientation="horizontal" class="v-divider theme--light">
    <div class="v-subheader sub-header font-weight-bold theme--light">
      Military
    </div>
    <div class="row">
      <div class="col-sm-1 col-lg-4 col-4">
        <div class="v-input theme--light v-text-field v-file-input">
          <div class="v-input__prepend-outer">
            <div class="v-input__icon v-input__icon--prepend"><button type="button" aria-label="prepend icon" class="v-icon notranslate v-icon--link mdi mdi-paperclip theme--light"></button></div>
          </div>
          <div class="v-input__control">
            <div class="v-input__slot">
              <div class="v-text-field__slot"><label for="input-20" class="v-label theme--light" style="left: 0px; position: absolute;">Military Document</label>
                <div class="v-file-input__text v-file-input__text--chips"></div><input accept=".pdf, .doc, .docx" id="input-20" type="file">
              </div>
              <div class="v-input__append-inner">
                <div></div>
              </div>
            </div>
            <div class="v-text-field__details">
              <div class="v-messages theme--light">
                <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <hr role="separator" aria-orientation="horizontal" class="v-divider theme--light">
    <div class="v-subheader sub-header font-weight-bold theme--light">
      Not born in US
    </div>
    <div class="row">
      <div class="col-sm-1 col-lg-4 col-4">
        <div class="v-input theme--light v-text-field v-file-input">
          <div class="v-input__prepend-outer">
            <div class="v-input__icon v-input__icon--prepend"><button type="button" aria-label="prepend icon" class="v-icon notranslate v-icon--link mdi mdi-paperclip theme--light"></button></div>
          </div>
          <div class="v-input__control">
            <div class="v-input__slot">
              <div class="v-text-field__slot"><label for="input-26" class="v-label theme--light" style="left: 0px; position: absolute;">Citizenship Documents</label>
                <div class="v-file-input__text v-file-input__text--chips"></div><input accept=".pdf, .doc, .dox" id="input-26" type="file">
              </div>
              <div class="v-input__append-inner">
                <div></div>
              </div>
            </div>
            <div class="v-text-field__details">
              <div class="v-messages theme--light">
                <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <hr role="separator" aria-orientation="horizontal" class="v-divider theme--light">
    <div class="v-subheader sub-header font-weight-bold theme--light">
      Supporting Documents
    </div>
    <div class="row">
      <div class="col-sm-1 col-lg-4 col-4">
        <div class="v-input theme--light v-text-field v-file-input">
          <div class="v-input__prepend-outer">
            <div class="v-input__icon v-input__icon--prepend"><button type="button" aria-label="prepend icon" class="v-icon notranslate v-icon--link mdi mdi-paperclip theme--light"></button></div>
          </div>
          <div class="v-input__control">
            <div class="v-input__slot">
              <div class="v-text-field__slot"><label for="input-32" class="v-label theme--light" style="left: 0px; position: absolute;">Supporting Documents</label>
                <div class="v-file-input__text v-file-input__text--chips"></div><input id="input-32" type="file" multiple="multiple">
              </div>
              <div class="v-input__append-inner">
                <div></div>
              </div>
            </div>
            <div class="v-text-field__details">
              <div class="v-messages theme--light">
                <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <hr role="separator" aria-orientation="horizontal" class="v-divider theme--light">
    <div class="v-subheader sub-header font-weight-bold theme--light">
      Legal name change
    </div>
    <div class="row">
      <div class="col-sm-1 col-lg-4 col-4">
        <div class="v-input theme--light v-text-field v-file-input">
          <div class="v-input__prepend-outer">
            <div class="v-input__icon v-input__icon--prepend"><button type="button" aria-label="prepend icon" class="v-icon notranslate v-icon--link mdi mdi-paperclip theme--light"></button></div>
          </div>
          <div class="v-input__control">
            <div class="v-input__slot">
              <div class="v-text-field__slot"><label for="input-38" class="v-label theme--light" style="left: 0px; position: absolute;">Name change documents</label>
                <div class="v-file-input__text v-file-input__text--chips"></div><input accept=".pdf, .doc, .dox" id="input-38" type="file">
              </div>
              <div class="v-input__append-inner">
                <div></div>
              </div>
            </div>
            <div class="v-text-field__details">
              <div class="v-messages theme--light">
                <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <hr role="separator" aria-orientation="horizontal" class="v-divider theme--light">
    <div class="v-subheader sub-header font-weight-bold theme--light">
      License Type Documents
    </div>
    <div class="row">
      <div class="col-sm-1 col-lg-4 col-4">
        <div class="v-input theme--light v-text-field v-file-input">
          <div class="v-input__prepend-outer">
            <div class="v-input__icon v-input__icon--prepend"><button type="button" aria-label="prepend icon" class="v-icon notranslate v-icon--link mdi mdi-paperclip theme--light"></button></div>
          </div>
          <div class="v-input__control">
            <div class="v-input__slot">
              <div class="v-text-field__slot"><label for="input-44" class="v-label theme--light" style="left: 0px; position: absolute;">Judicial documents</label>
                <div class="v-file-input__text v-file-input__text--chips"></div><input accept=".pdf, .doc, .dox" id="input-44" type="file">
              </div>
              <div class="v-input__append-inner">
                <div></div>
              </div>
            </div>
            <div class="v-text-field__details">
              <div class="v-messages theme--light">
                <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div class="col-sm-1 col-lg-4 col-4">
        <div class="v-input theme--light v-text-field v-file-input">
          <div class="v-input__prepend-outer">
            <div class="v-input__icon v-input__icon--prepend"><button type="button" aria-label="prepend icon" class="v-icon notranslate v-icon--link mdi mdi-paperclip theme--light"></button></div>
          </div>
          <div class="v-input__control">
            <div class="v-input__slot">
              <div class="v-text-field__slot"><label for="input-48" class="v-label theme--light" style="left: 0px; position: absolute;">Reserve documents</label>
                <div class="v-file-input__text v-file-input__text--chips"></div><input accept=".pdf, .doc, .dox" id="input-48" type="file">
              </div>
              <div class="v-input__append-inner">
                <div></div>
              </div>
            </div>
            <div class="v-text-field__details">
              <div class="v-messages theme--light">
                <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <hr role="separator" aria-orientation="horizontal" class="v-divider theme--light">
  </form>
  <div class="form-btn-container"><button type="button" class="v-btn v-btn--is-elevated v-btn--has-bg theme--light v-size--default success "><span class="v-btn__content">
    Continue
  </span></button> <button type="button" class="v-btn v-btn--is-elevated v-btn--has-bg theme--light v-size--default secondary mx-2"><span class="v-btn__content">
    Save and Exit
  </span></button> <button type="button" class="v-btn v-btn--is-elevated v-btn--has-bg theme--light v-size--default error mr-2"><span class="v-btn__content">
    Cancel
  </span></button></div>
  <div class="v-snack v-snack--bottom v-snack--has-background">
    <transition-stub name="v-snack-transition">
      <div class="v-snack__wrapper v-sheet theme--dark" style="display: none;">
        <div role="status" aria-live="polite" class="v-snack__content">
          error message
        </div>
        <div class="v-snack__action "></div>
      </div>
    </transition-stub>
  </div>
</div>
`;
